src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+        self.discard_ratio.store(ratio, Ordering::Relaxed);
+    }
+
+    fn adjust_memtables(&mut self, cf: &str) {
+        let num_memtables = self
+            .engine
+            .get_cf_num_immutable_mem_table(cf)
+            .unwrap_or(None)
+            .unwrap_or(0);
+        let checker = self.cf_checkers.get_mut(cf).unwrap();
+        SCHED_MEMTABLE_GAUGE
+            .with_label_values(&[cf])
+            .set(num_memtables as i64);
+        let prev = checker.last_num_memtables.get_recent();
+        checker.last_num_memtables.observe(num_memtables);
+
+        // do special check on start, see the comment of the variable definition for detail.
+        if checker.on_start_memtable {
+            if num_memtables < self.memtables_threshold
+                || checker.last_num_memtables.trend() == Trend::Increasing
+            {
+                // the write is accumulating, still need to throttle
+                checker.on_start_memtable = false;
+            } else {
+                // still on start, should not throttle now
+                return;
+            }
+        }
+
+        for c in self.cf_checkers.values() {
+            if num_memtables < c.last_num_memtables.get_recent() {
+                return;
+            }
+        }
+
+        let checker = self.cf_checkers.get_mut(cf).unwrap();
+        let is_throttled = self.limiter.speed_limit() != INFINITY;
+        let should_throttle =
+            checker.last_num_memtables.get_avg() > self.memtables_threshold as f64;
+        let throttle = if !is_throttled {
+            if should_throttle {
+                SCHED_THROTTLE_ACTION_COUNTER
+                    .with_label_values(&[cf, "memtable_init"])
+                    .inc();
+                checker.init_speed = true;
+                let x = self.write_flow_recorder.get_percentile_90();
+                if x == 0 { INFINITY } else { x as f64 }
+            } else {
+                INFINITY
+            }
+        } else if !should_throttle
+            || checker.last_num_memtables.get_recent() < self.memtables_threshold
+        {
+            // should not throttle memtable
+            checker.memtable_debt = 0.0;
+            if checker.init_speed {
+                INFINITY
+            } else {
+                self.limiter.speed_limit() + checker.memtable_debt
+            }
+        } else {
+            // should throttle
+            let diff = match checker.last_num_memtables.get_recent().cmp(&prev) {
+                std::cmp::Ordering::Greater => {
+                    checker.memtable_debt += 1.0;
+                    -1.0
+                }
+                std::cmp::Ordering::Less => {
+                    checker.memtable_debt -= 1.0;
+                    1.0
+                }
+                std::cmp::Ordering::Equal => {
+                    // keep, do nothing
+                    0.0
+                }
+            };
+            self.limiter.speed_limit() + diff * 1024.0 * 1024.0
```

Comment:
I find it a little weird to not consider memtable size in this algorithm.

Reply:
but how can you decide the rate by memtable size?

Reply:
No sure, consider when write buffer size is super small, the throttle should be stronger.